Support binary strings, preserve UTF-8 and UTF-16 errors

Maxdamantus opened the pull request:
The internal string representation is changed from UTF-8 with replacement characters to a modified form of "WTF-8" that is able to distinctly encode UTF-8 errors and UTF-16 errors.

This handles UTF-8 errors in raw string inputs and handles UTF-8 and UTF-16 errors in JSON input. UTF-16 errors (using "\uXXXX") and UTF-8 errors (using the original raw bytes) are maintained when emitting JSON. When emitting raw strings, UTF-8 errors are maintained and UTF-16 errors are converted into replacement characters.

As well as being able to be used as a "pretty printer" without risk of corrupting binary data (eg, JSON strings such as "\uDC00\uD800!", or arbitrary bytes in input), it is now possible to pass around arbitrary files either as JSON or raw data (using ill-formed UTF-8):

```
$ sha1sum jq
ff8e6e9fd7d87eb1c9179da546ccbbcd77a40f14  jq
$ ./jq -n '$a' --rawfile a jq | ./jq -j | sha1sum
ff8e6e9fd7d87eb1c9179da546ccbbcd77a40f14  -
$ base64 jq | ./jq -jR '@base64d' | sha1sum
ff8e6e9fd7d87eb1c9179da546ccbbcd77a40f14  -
```

To demonstrate UTF-16 error preservation:

```
$ node -e 'console.log(JSON.stringify("💩".split("")));'
["\ud83d","\udca9"]
$ node -e 'console.log(JSON.stringify("💩".split("")));' | ./jq .
[
  "\ud83d",
  "\udca9"
]
```

Fixes at least #1931, #1540, #2259.

Maxdamantus committed:
jv_string_implode: avoid producing unprintable string from reserved code points

Maxdamantus committed:
Update `@base64`, `utf8bytelength` and `fromjson` to handle binary strings

Maxdamantus committed:
Correct UTF-8 and UTF-16 errors during concatenation

UTF-8 errors and UTF-16 errors that were previously encoded into the ends of
strings will now potentially be used to form correct code points.

This is mostly a matter of making string equality behave expectedly, since
without this normalisation, it is possible to produce `jv` strings that are
converted to UTF-8 or UTF-16 the same way but are not equal due well-formed
code units that may or may not be encoded as errors.

nicowilliams commented:
@Maxdamantus sorry for the delay.  I've started to look at this, but in parallel I've been re-thinking how to do binary and have started to implement something not unlike your PR just to experiment.

First, my immediate thoughts about this PR:

 - jq should not be the first thing sporting "WTF-8b" support -- is it your invention?  are there other uses of it?
 - I'm not sure that we should steal from the over-long UTF-8 sequences to represent `0x80`-`0xff` -- like you I want UTF-16 to die and go away and never come back, and then we could have more than 21 bits of code space for Unicode, so using overlong sequences would not be advisable if I had a magic wand with which to get my way!
   Maybe we could use private-use codepoints instead?  Or the _last_ 128 overlong sequences rather than the first?
 - I'm very surprised to see changes to `src/builtin.c`:`f_json_parse()`!
   Why not just add new functions for new options?

Having binary be represented as an array of small numbers in jq code is probably a bad idea -- arrays are everywhere in jq, and adding all sorts of branches for binary seems like not likely to be good for performance.  String operations, I think, are probably less performance sensitive.  As well, I'm a bit disappointed with `explode` not producing a stream.  So here's my new idea, which I think we can use to add something like WTF-8 and so on like you're doing, possibly sharing code:

 - represent binary as a sub-kind of string
   - rename the `pad_` field of `jv` to `subkind`, and add a enum type for its values with a `JV_SUBKIND_BINARY`
   - a `jv` that is of kind `JV_KIND_STRING` with the sub-kind `JV_SUBKIND_BINARY` would be a binary string value
   - add `jv_binary_sized()` and related
   - make all the `jv` functions like `jv_string_concat()`, `jv_string_slice()`, `jv_string_append*()` that deal with strings support binary strings (meaning: don't check for UTF-8 validity and don't replace bad codepoints)
   - adding a string and a binary string (or a binary string and a string) should yield a binary string
 - maybe maybe maybe let `.[]` be the operation that explodes a string into its codepoints as a stream
   - then let `.[]` on a binary string output the string's _byte_ values, not _codepoints_
 - add `tobinary` for "converting" strings to binary, and make `tostring` on binary strings do bad codepoint replacement as needed
 - add a `subtype` function?
 - add `isbinary` to indicate whether a string is binary
 - add `isutf8` to indicate whether a string (possibly binary) is valid UTF-8
 - make `input_filename` output binary when the filename is not valid UTF-8
 - JSON output should by default do bad codepoint replacements as needed on binary values then output them as strings
   - if you want to have binary base64-encoded, use the base64 encoder
   - but there could be an output mode that outputs binary values as if they were valid UTF-8 strings
 - add command-line options for new raw input modes:
   - newline-terminated binary (so, text, but not necessarily UTF-8)
   - binary with some record size
   - binary slurp (read the whole file as one big binary string)
 - add command-line output modes:
   - raw binary output (no codepoint replacements on top-level strings)
   - raw binary output with a separator byte (ditto; see discussions on `-0`)
   - WTF-8 output (so, JSON, but any strings which are "valid WTF-8" would get treated as if they were valid UTF-8
   - binary output (JSON with any binary strings treated as if they were valid UTF-8
   - maybe a WTF-8b output mode?

nicowilliams commented:
@Maxdamantus see #2736.  What do you think?

Maxdamantus commented:
@nicowilliams Thanks for looking, better late than never!

>     * jq should not be the first thing sporting "WTF-8b" support -- is it your invention?  are there other uses of it?

As far as I can tell, noone else has used this particular scheme, but it's hard to verify. I've referred to it as "WTF-8b" because it achieves the goals of both WTF-8 (able to encode UTF-16 errors) and UTF-8b (able to encode UTF-8 errors, probably better known as "surrogateescape"). I've encoded UTF-16 errors the same way as in WTF-8, but it's not possible to directly use the UTF-8b/surrogateescape encoding in conjunction with WTF-8.

I think the use case here is a bit niche. I'm not aware of other systems where it would be appropriate to support both UTF-8 and UTF-16 strings at the same time.

IMO `jq` is should at least handle JSON produced by UTF-16 systems (eg, JavaScript where strings such as `"\udca9\ud83d"` can be emitted).

Since `jq` is commonly used as a unix tool for passing around text, I think it should also handle text in ways that are compatible with other unix tools. Eg, if I run `tail -n 20 foo.txt`, I expect it to maintain the same bytes that were in the original file—it shouldn't fail due to illegal formatting or replace some of the bytes with replacement characters.

The closest example of something like this combined encoding I can think of is Java's "Modified UTF-8". This is used in compiled `.class` files for representing string constants. UTF-16 data is encoded mostly[0] the same way as WTF-8, but instead of encoding `"\u0000"` as a single byte `"\x00"`, it's encoded using the first overlong coding, `"\xC0\x80"`.

I'll also point out WTF-8b is intended as an internal encoding. In the current changeset, it is exposed through `@uri`, but I think it would be cleaner to transform the UTF-16 errors here into replacement characters. I have however found `@uri` useful for inspecting the internal representation. Theoretically it should be possible to change the internal representation from WTF-8b to something else (though I'm doubtful that a better encoding exists).

>     * I'm not sure that we should steal from the over-long UTF-8 sequences to represent `0x80`-`0xff` -- like you I want UTF-16 to die and go away and never come back, and then we could have more than 21 bits of code space for Unicode, so using overlong sequences would not be advisable if I had a magic wand with which to get my way!
>       Maybe we could use private-use codepoints instead?  Or the _last_ 128 overlong sequences rather than the first?

I don't think private-use code points would be suitable, as these are valid Unicode, so it would be incorrect to reinterpret them as UTF-8 errors (or to interpret UTF-8 errors as valid Unicode), eg [HERE](https://github.com/oniani/covid-19-chatbot/blob/46f0afd5341b7f8061779564125d7ca5481c5b10/data_raw/1f4ec41f723e758522faa99829a52f00ea45a9e2.json#L2501) 's something I found through GitHub code search, where there are private-use code points that presumably map to characters in a special font. These characters need to be distinguished from UTF-8 errors.

I'm not sure what the advantage would be of using the last 128 overlong sequences rather than the first ones. The last overlong sequences would be 4 bytes long, and there are 65,536 overlong 4-byte sequences (ie, the number of code points below `U+10000`, the first that is encoded as 4 bytes in UTF-8).

Using the first overlong sequences seems quite natural, since they are only 2 bytes long, and there are exactly 128 overlong 2-byte sequences, and exactly 128 possible unpaired UTF-8 bytes. It seems to match exactly, so if someone else were to independently create an encoding under the same constraints, they'd likely come up with the same mapping.

>     * I'm very surprised to see changes to `src/builtin.c`:`f_json_parse()`!
>       Why not just add new functions for new options?

I think my intention was to avoid complicating the existing `jv_parse_sized` function or to avoid creating an extra exposed function (`jv_parse_extended_sized`?). I believe this was the part of the PR that I felt least comfortable with stylistically, so I'd be happy to revisit this.

> So here's my new idea, which I think we can use to add something like WTF-8 and so on like you're doing, possibly sharing code:

I certainly think further operations on binary strings would be useful (my PR is not focused on providing such functions, but only adding representational support so that the data does not get corrupted or cause errors), but I feel like creating hard distinctions between different types of strings is unnecessary.

I think the main thing you'd want from "binary" strings is byte-based indexing instead of code point-based indexing.

`jq` currently does neither, but I see that your PR (#2736) is working towards code point-based indexing for normal strings, which makes sense (actually, IIUC, it only does iterating at the moment, but I think indexing is a logical next step, so that `.[$x]` is the same as `[.[]][$x]`).

Have you thought about making binary strings that are basically just alternative views over normal strings? So `$str | .[]` would emit code points[1], but `$str | asbinary | .[]` would emit bytes. `$str | asbinary | asstring` could just be a noop (it could force the UTF-16 errors into replacement characters, but I don't think it's very useful). If the indexing is meant to return a string (or binary string) instead of an integer, I think it would still be sensible for `"\uD800 X"[0]` and `("\uD800 X" | asbinary)[0]` to return the same string (or binary string) representing that illegal UTF-16 surrogate (this adheres to the principle of preserving data, but in practice I think this is an extreme edge case, since you wouldn't normally perform binary operations on ill-formed UTF-16 data).

I suspect the alternative view method would be simpler implementation-wise, since having different string/binary representations probably means the C code will have to cater for the different representations (eg, `@base64` is going to have to convert from normal strings and binary strings .. and what type does `@base64d` emit? at the moment it emits a normal string, which should iterate through the code points—changing it to return a binary string seems a bit presumtive).

[0] Actually, it's slightly different in that it treats all surrogates equally as errors, whether they are paired or not, so `"💩"` is first turned into the UTF-16 surrogates `<D83D DCA9>` and the surrogates are encoded individually.

[1] Not sure if these should be strings or integers. It looks like your PR makes them strings. I don't think I feel strongly either way, but I have a slight tendendency in favour of `("💩" | asbinary | .[0]) == 240`, so for consistency I might expect `("💩" | .[0]) == 128169`. I think in both cases, `.[0:(. | length)] == .` is reasonable (a slice will always produce the same type as the input).

Maxdamantus committed:
Binary strings: preserve UTF-8 and UTF-16 errors

The internal string representation is changed from UTF-8 with replacement
characters to a modified form of "WTF-8" that is able to distinctly encode
UTF-8 errors and UTF-16 errors.

This handles UTF-8 errors in raw string inputs and handles UTF-8 and UTF-16
errors in JSON input. UTF-16 errors (using "\uXXXX") and UTF-8 errors (using
the original raw bytes) are maintained when emitting JSON. When emitting raw
strings, UTF-8 errors are maintained and UTF-16 errors are converted into
replacement characters.

Maxdamantus committed:
Update `@uri` to handle binary strings

ikegami commented:
> I don't think private-use code points would be suitable, as these are
valid Unicode

Indeed.

> where there are private-use code points that presumably map to characters
in a special font

There's a group that assigned Klingon characters to a range of private-use
code points.


On Thu, Jul 20, 2023 at 6:03 AM Maxdamantus ***@***.***>
wrote:

> @nicowilliams <https://github.com/nicowilliams> Thanks for looking,
> better late than never!
>
> * jq should not be the first thing sporting "WTF-8b" support -- is it your invention?  are there other uses of it?
>
> As far as I can tell, noone else has used this particular scheme, but it's
> hard to verify. I've referred to it as "WTF-8b" because it achieves the
> goals of both WTF-8 (able to encode UTF-16 errors) and UTF-8b (able to
> encode UTF-8 errors, probably better known as "surrogateescape"). I've
> encoded UTF-16 errors the same way as in WTF-8, but it's not possible to
> directly use the UTF-8b/surrogateescape encoding in conjunction with WTF-8.
>
> I think the use case here is a bit niche. I'm not aware of other systems
> where it would be appropriate to support both UTF-8 and UTF-16 strings at
> the same time.
>
> IMO jq is should at least handle JSON produced by UTF-16 systems (eg,
> JavaScript where strings such as "\udca9\ud83d" can be emitted).
>
> Since jq is commonly used as a unix tool for passing around text, I think
> it should also handle text in ways that are compatible with other unix
> tools. Eg, if I run tail -n foo.txt, I expect it to maintain the same
> bytes that were in the original file—it shouldn't fail due to illegal
> formatting or replace some of the bytes with replacement characters.
>
> The closest example of something like this combined encoding I can think
> of is Java's "Modified UTF-8". This is used in compiled .class files for
> representing string constants. UTF-16 data is encoded mostly[0] the same
> way as WTF-8, but instead of encoding "\u0000" as a single byte "\x00",
> it's encoded using the first overlong coding, "\xC0\x80".
>
> I'll also point out WTF-8b is intended as an internal encoding. In the
> current changeset, it is exposed through @uri, but I think it would be
> cleaner to transform the UTF-16 errors here into replacement characters. I
> have however found @uri useful for inspecting the internal
> representation. Theoretically it should be possible to change the internal
> representation from WTF-8b to something else (though I'm doubtful that a
> better encoding exists).
>
> * I'm not sure that we should steal from the over-long UTF-8 sequences to represent `0x80`-`0xff` -- like you I want UTF-16 to die and go away and never come back, and then we could have more than 21 bits of code space for Unicode, so using overlong sequences would not be advisable if I had a magic wand with which to get my way!
>   Maybe we could use private-use codepoints instead?  Or the _last_ 128 overlong sequences rather than the first?
>
> I don't think private-use code points would be suitable, as these are
> valid Unicode, so it would be incorrect to reinterpret them as UTF-8 errors
> (or to interpret UTF-8 errors as valid Unicode), eg HERE
> <https://github.com/oniani/covid-19-chatbot/blob/46f0afd5341b7f8061779564125d7ca5481c5b10/data_raw/1f4ec41f723e758522faa99829a52f00ea45a9e2.json#L2501>
> 's something I found through GitHub code search, where there are
> private-use code points that presumably map to characters in a special
> font. These characters need to be distinguished from UTF-8 errors.
>
> I'm not sure what the advantage would be of using the last 128 overlong
> sequences rather than the first ones. The last overlong sequences would be
> 4 bytes long, and there are 65,536 overlong 4-byte sequences (ie, the
> number of code points below U+10000, the first that is encoded as 4 bytes
> in UTF-8).
>
> Using the first overlong sequences seems quite natural, since they are
> only 2 bytes long, and there are exactly 128 overlong 2-byte sequences, and
> exactly 128 possible unpaired UTF-8 bytes. It seems to match exactly, so if
> someone else were to independently create an encoding under the same
> constraints, they'd likely come up with the same mapping.
>
> * I'm very surprised to see changes to `src/builtin.c`:`f_json_parse()`!
>   Why not just add new functions for new options?
>
> I think my intention was to avoid complicating the existing jv_parse_sized
> function or to avoid creating an extra exposed function (
> jv_parse_extended_sized?). I believe this was the part of the PR that I
> felt least comfortable with stylistically, so I'd be happy to revisit this.
>
> So here's my new idea, which I think we can use to add something like
> WTF-8 and so on like you're doing, possibly sharing code:
>
> I certainly think further operations on binary strings would be useful (my
> PR is not focused on providing such functions, but only adding
> representational support so that the data does not get corrupted or cause
> errors), but I feel like creating hard distinctions between different types
> of strings is unnecessary.
>
> I think the main thing you'd want from "binary" strings is byte-based
> indexing instead of code point-based indexing.
>
> jq currently does neither, but I see that your PR (#2736
> <https://github.com/jqlang/jq/pull/2736>) is working towards code
> point-based indexing for normal strings, which makes sense (actually, IIUC,
> it only does iterating at the moment, but I think indexing is a logical
> next step, so that .[$x] is the same as [.[]][$x]).
>
> Have you thought about making binary strings that are basically just
> alternative views over normal strings? So $str | .[] would emit code
> points[1], but $str | asbinary | .[] would emit bytes. $str | asbinary |
> asstring could just be a noop (it could force the UTF-16 errors into
> replacement characters, but I don't think it's very useful). If the
> indexing is meant to return a string (or binary string) instead of an
> integer, I think it would still be sensible for "\uD800 X"[0] and ("\uD800
> X" | asbinary)[0] to return the same string (or binary string)
> representing that illegal UTF-16 surrogate (this adheres to the principle
> of preserving data, but in practice I think this is an extreme edge case,
> since you wouldn't normally perform binary operations on ill-formed UTF-16
> data).
>
> I suspect the alternative view method would be simpler
> implementation-wise, since having different string/binary representations
> probably means the C code will have to cater for the different
> representations (eg, @base64 is going to have to convert from normal
> strings and binary strings .. and what type does @base64d emit? at the
> moment it emits a normal string, which should iterate through the code
> points—changing it to return a binary string seems a bit presumtive).
>
> [0] Actually, it's slightly different in that it treats all surrogates
> equally as errors, whether they are paired or not, so "💩" is first
> turned into the UTF-16 surrogates <D83D DCA9> and the surrogates are
> encoded individually.
>
> [1] Not sure if these should be strings or integers. It looks like your PR
> makes them strings. I don't think I feel strongly either way, but for
> consistency between normal strings and binary strings, it might make more
> sense
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/jqlang/jq/pull/2314#issuecomment-1643551881>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFK2ZPG6ZQXNP76LB3J5XTXRDX7LANCNFSM45GTTNYQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

Maxdamantus commented:
I've rebased this PR onto the current master without making any significant changes (only resolving a merge conflict with #2633). I intend to push some code improvements later.

nicowilliams commented:
> > I don't think private-use code points would be suitable, as these are

> valid Unicode Indeed.

But on the plus side such output would be accepted and preserved by existing software.

> > where there are private-use code points that presumably map to characters

> in a special font There's a group that assigned Klingon characters to a range of private-use code points.

I'm aware.  Ideally we could get 128 codepoints assigned for this.  Yes, you'd have to know to decode from UTF-8 using those to binary, but their presence would be sufficient to indicate that a string is in fact a binary blob.

Maxdamantus commented:
> Yes, you'd have to know to decode from UTF-8 using those to binary, but their presence would be sufficient to indicate that a string is in fact a binary blob.

You would need an external flag in this case to denote that it's a binary blob, since it's possible for binary data to already contain bytes that look like an encoding of these hypothetical code points.

For example, imagine that `U+E1XX` is reserved for this purpose, where `U+E180` would denote the binary byte `<80>`. If some binary input happens to contain `<EE 86 80>` (the UTF-8 encoding of `U+E180`), that will presumably get decoded as `<80>` on output, corrupting the binary data.

If an external flag is used, there isn't much point in reencoding the binary data, since it could be stored in plain binary form. The downside is that operations have to account for the different string representations (and concatenation between different representations would fail in some cases).

nicowilliams commented:
> > Yes, you'd have to know to decode from UTF-8 using those to binary, but their presence would be sufficient to indicate that a string is in fact a binary blob.
> 
> You would need an external flag in this case to denote that it's a binary blob, since it's possible for binary data to already contain bytes that look like an encoding of these hypothetical code points.

You'd have to know yes, and the point is that if _jq_ produces it (unless it's a mode where it produces actual binary) then it's UTF-8.  The "external flag" here is implied as long as you don't leak this "UTF-8" into contexts where binary is expected without first decoding.

We have this problem in spades.  Filesystem related system calls in Unix are all 8-bit clean except for ASCII `/` and ASCII `NUL`, and so if you list a directory you've no idea what codeset each name is in unless you enforce a convention.

Anyways, I think for jq all of this is out of scope.  WTF-8 might be in scope, but WTF-8b... I think would require socializing more.

Maxdamantus committed:
Preserve UTF-8 and UTF-16 errors in `explode`

Errors are emitted as negative code points instead of being transformed into
replacement characters. `implode` is also updated accordingly so the original
string can be reconstructed without data loss.

Maxdamantus committed:
Remove UTF-8 backtracking workaround

This is no longer needed as strings are capable of storing partial UTF-8
sequences.

Maxdamantus commented:
I've made some updates to these changes:

- `f_json_parse` reuses the parse function in `jv_parse.c` again, which has been refactored to work on both types of strings.
- Various functions that I had previously called `*_extended_*` are now called `*_wtf_*`, since the intention is to denote that they work with the extended (internal) string representation, which happens to be WTF-8/WTF-8b. The previous names were more abstract (not referring to implementation details of the representation), but I think the new names are clearer.
- A `jvp_utf8_wtf_next_bytes` function has been added which makes iteration through UTF-8 bytes simpler, where for well-formed strings, the entire content will be emitted in one chunk.
- `explode` and `implode` preserve errors by using negative code points. I was thinking of having separate functions that do this, but given that until these changes, binary data hasn't really worked properly, it should be compatible with current usage of `jq`.
- `@uri` is implemented properly, so the internal representation is no longer exposed for ill-formed strings.
- The `jvp_utf8_backtrack` function is removed, since this was a workaround for the fact that strings couldn't hold partial UTF-8 data, but now they can.

Maxdamantus committed:
tests/shtest: fix use of base64 command for macOS compatibility
